palette/src/js/script.js

```diff
@@ -0,0 +1,409 @@
+    this.prevColor = this.currentColor;
+    this.currentColor = newColor;
+    document.querySelector('#current-color i').style.color = this.currentColor;
+    document.querySelector('#prev-color i').style.color = this.prevColor;
+    this.ctx.fillStyle = this.currentColor;
+    document.getElementById('inp-color').value = this.currentColor;
+  }
+
+  // get position in ImageData
+  getPixelPos(x, y) {
+    return (y * this.scale + x) * 4;
+  }
+
+  // return Star Color
+  static matchStartColor(data, pos, startColor) {
+    return (data[pos] === startColor.r
+      && data[pos + 1] === startColor.g
+      && data[pos + 2] === startColor.b
+      && data[pos + 3] === startColor.a);
+  }
+
+  // set pixel color
+  static colorPixel(dat, pos, color) {
+    const data = dat;
+    data[pos] = color.r || 0;
+    data[pos + 1] = color.g || 0;
+    data[pos + 2] = color.b || 0;
+    data[pos + 3] = Object.prototype.hasOwnProperty.call(color, 'a') ? color.a : 255;
+  }
+
+  // convert color format
+  fillPromise(data) {
+    const promiseFill = new Promise((resolve) => {
+      if (this.fillArea(data)) {
+        setTimeout(() => {
+          resolve();
+        }, 0);
+      }
+    });
+    promiseFill.then(() => {
+      this.busy = false;
+    });
+  }
+
+  // convert format color
+  static colorToRgba(c) {
+    let result;
+    if (c[0] === '#') {
+      const color = c.replace('#', '');
+      const bigint = parseInt(color, 16);
+      // eslint-disable-next-line no-bitwise
+      const r = (bigint >> 16) & 255;
+      // eslint-disable-next-line no-bitwise
+      const g = (bigint >> 8) & 255;
+      // eslint-disable-next-line no-bitwise
+      const b = bigint & 255;
+      result = {
+        r,
+        g,
+        b,
+        a: 255,
+      };
+    } else if (c.indexOf('rgba(') === 0) {
+      const color = c.replace('rgba(', '').replace(' ', '').replace(')', '').split(',');
+      result = {
+        r: color[0],
+        g: color[1],
+        b: color[2],
+        a: color[3] * 255,
+      };
+    } else {
+      result = {
+        r: 0,
+        g: 0,
+        b: 0,
+        a: 0,
+      };
+    }
+    return result;
+  }
+
+  saveLocalStorage() {
+    if (localStorage.getItem('state')) {
+      localStorage.removeItem('paletteLayout');
+      localStorage.removeItem('state');
+      const state = {
+        mode: this.mode,
+        currentColor: this.currentColor,
+        prevColor: this.prevColor,
+      };
+      localStorage.setItem('paletteLayout', this.canvas.toDataURL());
+      localStorage.setItem('state', JSON.stringify(state));
+    } else {
+      const state = {
+        mode: this.mode,
+        currentColor: this.currentColor,
+        prevColor: this.prevColor,
+      };
+      localStorage.setItem('state', JSON.stringify(state));
+    }
+  }
+
+  cleanCanvas() {
+    this.ctx.clearRect(0, 0, this.scale, this.scale);
+  }
+}
+
+const palette = new Palette();
+document.body.addEventListener('load', palette.bodyLoad());
```

Comment:
the class `Palette` could be in a separate file and the run logic - in the main one. Don't be afraid of breaking down the app into several files, it would simplify code maintenance in the future